pkgdb/include/flox/pkgdb/write.hh

```diff
@@ -30,6 +34,117 @@ using Target = std::tuple<flox::AttrPath, flox::Cursor, row_id>;
+
+/** @brief Scraping rules to modify database creation process in _raw_ form. */
+struct ScrapeRulesRaw
+{
+  std::vector<AttrPathGlob> allowPackage;
+  std::vector<AttrPathGlob> disallowPackage;
+  std::vector<AttrPathGlob> allowRecursive;
+  std::vector<AttrPathGlob> disallowRecursive;
+}; /* End struct `ScrapeRulesRaw` */
+
+
+/** @brief Convert a JSON object to a @a flox::pkgdb::ScrapeRulesRaw. */
+void
+from_json( const nlohmann::json & jfrom, ScrapeRulesRaw & rules );
+
+
+/* -------------------------------------------------------------------------- */
+
+enum ScrapeRule {
+  SR_NONE = 0,         /**< Empty state. */
+  SR_DEFAULT,          /**< Applies no special rules. */
+  SR_ALLOW_PACKAGE,    /**< Forces an package entry in DB. */
+  SR_ALLOW_RECURSIVE,  /**< Forces a sub-tree to be scraped. */
+  SR_DISALLOW_PACKAGE, /**< Do not add package entry to DB. */
+  /** Ignore sub-tree members unless otherwise specified. */
+  SR_DISALLOW_RECURSIVE
+}; /* End enum `ScrapeRule` */
+
+[[nodiscard]] std::string
+scrapeRuleToString( ScrapeRule rule );
+
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Node definition for a rules tree.
+ *
+ * The tree is built with a root node, where each node contains an attribute
+ * name, and the rule to be applied, along with a list of child nodes.
+ * This tree is built from reading the rules file, with paths through the tree
+ * constucted with SR_DEFAULT rules along the path until a leaf node with the
+ * appropriate rule can be added.  This allows hierarchical searching through
+ * the tree for attribute paths encountered during scraping and maintains the
+ * context for child inheritance of the rule defined for the deepest ancestral
+ * node.
+ */
+struct RulesTreeNode
+{
+  using Children = std::unordered_map<std::string, RulesTreeNode>;
+
+  std::string attrName = "";
+  ScrapeRule  rule     = SR_DEFAULT;
+  Children    children = {};
+
+  RulesTreeNode() = default;
+
+  explicit RulesTreeNode( ScrapeRulesRaw rules );
+
+  explicit RulesTreeNode( const std::filesystem::path & path )
+    : RulesTreeNode( static_cast<ScrapeRulesRaw>( readAndCoerceJSON( path ) ) )
+  {}
+
+  explicit RulesTreeNode( std::string attrName,
+                          ScrapeRule  rule     = SR_DEFAULT,
+                          Children    children = {} )
+    : attrName( std::move( attrName ) )
+    , rule( std::move( rule ) )
+    , children( std::move( children ) )
+  {}
+
+  RulesTreeNode( std::string attrName, Children children )
+    : attrName( std::move( attrName ) ), children( std::move( children ) )
+  {}
+
+  void
+  addRule( AttrPathGlob & relPath, ScrapeRule rule );
+
+  /**
+   * @brief Get the rule at a path, or @a flox::pkgdb::SR_DEFAULT as a fallback.
+   *
+   * This *does NOT* apply parent rules to children.
+   *
+   * @see @a flox::pkgdb::RulesTreeNode::applyRules
+   */
+  [[nodiscard]] ScrapeRule
+  getRule( const AttrPath & path = {} ) const;
```

Comment:
Question: similar question here, what is the relationship between this node and the attrpath that's being provided?